cupyx/jit/_compile.py

```diff
@@ -172,19 +191,43 @@ def _eval_operand(op, args, env, dtype=None):
     if op is None:
         raise TypeError(
             f'"{ufunc.name}" does not support for the input types: {in_types}')
-    if op.routine is None:
-        op.error_func()
 
-    assert op.routine.startswith('out0 = ')
-    out_type = _types.Scalar(op.out_types[0])
-    expr = op.routine[7:]
-    for i, x in enumerate(args):
-        x = astype_scalar(x, _types.Scalar(op.in_types[i]))
-        expr = expr.replace(f'in{i}', x.code)
-    expr = expr.replace('out0_type', str(out_type))
-    env.preambles.add(ufunc._preamble)
+    if op.error_func is not None:
+        op.error_func()
 
-    return CudaObject('(' + expr + ')', out_type)
+    if ufunc.nout == 1 and op.routine.startswith('out0 = '):
+        out_type = _types.Scalar(op.out_types[0])
+        expr = op.routine.replace('out0 = ', '')
+        args = [_astype_scalar(x, _types.Scalar(t))
+                for x, t in zip(args, op.in_types)]
+
+        can_use_inline_expansion = True
+        for i in range(ufunc.nin):
+            if len(list(re.finditer(r'in{}'.format(i), op.routine))) > 1:
+                can_use_inline_expansion = False
+
+        if can_use_inline_expansion:
+            # Code pass for readable generated code
+            for i, x in enumerate(args):
+                expr = expr.replace(f'in{i}', x.code)
+            expr = '(' + expr.replace('out0_type', str(out_type)) + ')'
+            env.preambles.add(ufunc._preamble)
+        else:
+            template_typenames = ', '.join([
+                f'typename T{i}' for i in range(ufunc.nin)])
+            ufunc_name = f'{ufunc.name}_{str(numpy.dtype(op.out_types[0]))}'
+            params = ', '.join([f'T{i} in{i}' for i in range(ufunc.nin)])
+            ufunc_code = f"""template <{template_typenames}>
+__device__ {out_type} {ufunc_name}({params}) {{
```

Comment:
Do we wanna inline it?
```suggestion
__device__ __forceinline__ {out_type} {ufunc_name}({params}) {{
```

Reply:
The inline expansion I commented on https://github.com/cupy/cupy/pull/4347#issuecomment-766612103 is for the readability of generated code, not for the performance improvement. Do you think `__forceinline__` is needed for the performance?

Reply:
I wonder the real effect of this.

Reply:
Let's go without it and trust the compiler's heuristics! 🙂